Use CLOCK_MONOTONIC for rate limiting (fix #706)

maxmouchet opened the pull request:
This replaces calls to `gettimeofday` with `clock_gettime(CLOCK_MONOTONIC, ...)` in the rate limiting code.
This fixes #706 where ZMap can crash if the host clock goes backwards.

Feel free to suggest a better name/location for the `steady_now` function!

zakird commented:
Overall this looks good, but reading the POSIX standard, it does seem like we do need to ensure support for `CLOCK_MONOTONIC` before we use it . Only `CLOCK_REALTIME` is guaranteed to exist on POSIX. (I would imagine that most systems we use support this, but we've been bitten a bunch of times in the past — people run ZMap on strange platforms.) If I'm reading the docs right, we can check whether `_POSIX_TIMERS` is defined to ensure `clock_gettime` is available. If that looks good, we also need to check that `_POSIX_MONOTONIC_CLOCK` also is also there.

> On POSIX systems on which these functions are available, the symbol _POSIX_TIMERS is defined in <[unistd.h](https://linux.die.net/include/unistd.h)> to a value greater than 0. The symbols _POSIX_MONOTONIC_CLOCK, _POSIX_CPUTIME, _POSIX_THREAD_CPUTIME indicate that CLOCK_MONOTONIC, CLOCK_PROCESS_CPUTIME_ID, CLOCK_THREAD_CPUTIME_ID are available

maxmouchet commented:
I've moved the function to `logger.c` and added a bit of documentation in `logger.h`.
I've also added the POSIX check, which seems to work as expected on my machine.
I thought about adding a `#pragma message(...)` directive to log which implementation is chosen but I don't know if it's necessary and/or portable across compilers.

zakird commented:
I moved the functions to `util.c` since they don't really belong in `logger.c`. Not your bad — I don't know why `now()` was defined there int the first place. I merged into main in #708.

maxmouchet commented:
Awesome, thanks!
